enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,473 @@
+  - "@enxebre"
+  - "@eparis"
+  - "@mrunalp"
+  - "@sttts"
+creation-date: 2020-08-31
+last-updated: 2020-10-08
+status: implementable
+replaces:
+superseded-by:
+---
+
+# Add support for out-of-tree cloud provider integration
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposal describes the migration of cloud platforms from the deprecated [in-tree cloud providers](https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/) to [Cloud Controller Manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/) services that implement `external cloud provider` [interface](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go).
+
+As a pioneer platform, it is proposed to use OpenStack.
+
+## Motivation
+
+Using Cloud Controller Managers (CCMs) is the Kubernetes' [preferred way](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/) to interact with underlying cloud platforms as it provides more flexibility and freedom for developers. It replaces existing in-tree cloud providers, which have been deprecated and will be permanently removed approximately in Kubernetes [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). But they are still used in OpenShift and we must start a smooth migration towards CCMs.
+
+Another motivation is to be closer to upstream by helping develop Cloud Controller Managers for various platforms, which is benefiting both OpenShift and Kubernetes.
+
+The change will help adding support for other cloud platforms, such as [Packet](https://github.com/packethost/packet-ccm), [Digital Ocean](https://github.com/digitalocean/digitalocean-cloud-controller-manager) or [Alibaba Cloud](https://github.com/kubernetes/cloud-provider-alibaba-cloud).
+
+It's especially important to do this for OpenStack because switching to the external cloud provider fixes many issues and limitations with the in-tree cloud provider, such as it's reliance on [Nova metadata service](https://docs.openstack.org/nova/latest/admin/metadata-service.html). For OpenStack platform, this means the possibility for deploying on provider networks and at the edge.
+
+### Goals
+
+- Prepare OpenShift components to accommodate the CCMs instead of deprecated in-tree cloud providers.
+- Provide the means to allow management of generic `CCM` component based on the platform specific upstream implementation.
+- Define and implement an upgrading and downgrading paths between the in-tree and the out-of-tree cloud provider configurations.
+- Ensure the transition will ensure the feature parity between the in-tree and the out-of-tree provider.
```

Comment:
What should this sentence mean? Why do they have to have feature parity?

Reply:
I think this could be (based on my interpretation of this statement - we don't want to lose any functionality):
```
- Ensure full feature parity between the in-tree and the out-of-tree providers before migration to ensure core Kubernetes cloud features (eg Load Balancer Services) continue to operate as before.
```
But I wonder if that may be a tad long winded for a Goal, WDYT?